Reworking 9.7.3 (#88)
bhoffman0/CSAwesome: _sources/Unit9-Inheritance/topic-9-7-Object.rst

Object Superclass
====================

The ``Object`` class is the superclass of all other classes in Java and a part
of the built-in ``java.lang`` package. If a parent class isn't specified using
the ``extends`` keyword, the class will inherit from the ``Object`` class. What
does a class inherit from the ``Object`` class? The |AP CSA Reference Sheet|

> lists the two main methods that are most frequenly:

Reviewer: ```suggestion
lists the two main methods that are most frequently used:
```

   ``o1.equals(o3)``.

#. Equality is **consistent**: ``o1.equals(o2)`` always returns the same value
   assuming the objects are not modified.

#. No object is equal to ``null``: ``o.equals(null)`` is always ``false``.

The other way to look at these requirements is as guarantees that are made to
you as a user of ``equals``. If you look at it that way, these requirements are
quite nice. Imagine how much harder it would be to use the ``String`` equals
method if you couldn't rely on the fact that ``s1.equals(s2)`` is necessarily
the same as ``s2.equals(s1)``!

So even though the Java compiler can't force you to implement ``equals``
correctly, if you ever do want to override it, it’s important that you do. Let’s
look at what’s involved.

We'll write a class ``Word`` which represents a word in a particular language.
We want two ``Word`` objects to be considered ``equals`` if and only if they are
spelled the same `and` come from the same languag. The latter requirement is
because sometimes different languages have words that are spelled the same but
with different meanings such as “pie” which in English is a tasty baked treat

> and in Spanish which is what we call a “foot” in English.

Reviewer: ```suggestion
and in Spanish is what we call a “foot” in English.
```

   :language: java
   :autograde: unittest

   Try to guess what this code will print out before running it. Click on the
   CodeLens button to step forward through the code and watch the memory.

   ~~~~

>    public class Word

Reviewer: Nice example!

   comparisons of the individual attributes since two objects should only be
   equal if `all` the attributes match.

Note that the requirements on ``equals`` make it almost impossible to correctly
override it in a subclass of a class that has already overridden the ``Object``
version. To see why, imagine if we made a subclass of ``Word``,
``ClassifiedWord`` and added another attribute, ``partOfSpeech``.

If we override ``equals`` in the ``ClassifiedWord`` to only consider two
``ClassifiedWord`` objects ``equals`` if their spelling, language, `and` part of
speech match, that will break the symmetry since
``regularWord.equals(classifiedWord)`` will invoke the ``equals`` from ``Word``
which will only compare the spelling and language of the word but
``classifiedWord.equals(regularWord)`` will return ``false`` assuming the
``equals`` in ``ClassifiedWord`` checks that ``other`` is an ``instanceof
ClassifiedWord``. In general you should only provide an overridden ``equals``
method in one class in a class hierarchy.

|Groupwork| Programming Challenge : Savings Account
---------------------------------------------------

> In the following code, contains the beginning of a class for representing a bank

Reviewer: ```suggestion
The following code contains the beginning of a class for representing a bank
```

         public void test1()
         {
             String output = getMethodOutput("main");

>              String expect = "Armani Smith, 1500.0\nDakota Jones, 1500.0, 4.5";

Reviewer: Is Armani Smith still printed out? I wasn't sure if you deleted that.

Author: It still is.
